hashicorp/dev-portal
# feat: adds themed image support

## 🔗 Relevant links



- [Preview link](https://dev-portal-git-ksadds-themed-image-support-hashicorp.vercel.app/) 🔎
- [Asana task](https://app.asana.com/0/1199634971449915/1204334498378837) 🎟️

## 🗒️ What

|  dark  | light  |  
|---|---|
|  <img width="979" alt="Screenshot 2023-06-23 at 10 45 40 AM" src="https://github.com/hashicorp/dev-portal/assets/36613477/9a852a2e-0c6d-4ed8-b961-65d40f544c5a"> |  <img width="983" alt="Screenshot 2023-06-23 at 10 45 47 AM" src="https://github.com/hashicorp/dev-portal/assets/36613477/e1274726-5553-45b4-a2fb-2dc8b4bedfd1"> |   

This PR adds support for conditionally hiding images based on theme states by appending a hash to the md source content:
```md
![dark theme](/img/themed/dark-test.png#hide-on-light)
```
The `rewriteStaticAssets` plugin was adjusted to support this hash, and a `getTheme` function is used within the `Image` component to determine whether the `data-hide-on-theme` attribute should be added. 

## 🧪 Testing

- [Visit the preview](https://dev-portal-git-ksadds-themed-image-support-hashicorp.vercel.app/well-architected-framework/operational-excellence/operational-excellence-introduction) where a stub image with the hash is in use via  [staging](https://github.com/hashicorp/tutorials/blob/staging/content/tutorials/well-architected-framework/operational-excellence-introduction.mdx#L15-L16). 
- A[nd another](https://dev-portal-git-ksadds-themed-image-support-hashicorp.vercel.app/well-architected-framework/com/cloud-operating-model) that has plain images aswell
- Switch themes, validate that the asset changes
- Check the markup, validate that two images are there with the `data-hide-on-theme` attribute set. 
- Validate that un-themed images are unaffected in tutorials content, docs content etc.

## 💭 Anything else?

In follow up work, I'll add a remark plugin that automatically detects width / height for images, I chose to separate that out from this work.
